### PR TITLE
#3 S3 support

### PR DESCRIPTION
**Pull request by sbcd90**
A simple example of using s3 instead of gcs.

cargo-fetcher.exe --url **** --lock-file test.txt sync
cargo-fetcher.exe --url **** --lock-file test.txt mirror

Please review.

### CODE REVIEW

**Comment by Jake-Shadle** [comment]
I've changed things in #10  so GCS is now behind the `gcs` feature, so S3 can be used instead of, or in parallel with GCS and should be easier to add even more backends.

**Comment by sbcd90** [comment]
Hi @Jake-Shadle , I have updated the PR based on your latest changes.
I'm still using `rusoto` as it seems to be the most popular & well maintained rust library for aws.

I didnt modify `deny.toml`. Hence, `cargo-deny` check fails for now.
Will wait for your final decision.

File: src/sync.rs [hunk 1]
```diff
@@ -81,9 +81,9 @@ pub fn locked_crates(ctx: &crate::Ctx<'_>, root_dir: &Path) -> Result<(), Error>
         use std::fmt::Write;
         write!(&mut krate_name, "{}", krate.local_id()).unwrap();
 
-        if cached_crates.binary_search(&krate_name).is_err() {
-            to_sync.push(krate);
-        }
+        //        if cached_crates.binary_search(&krate_name).is_err() {
+        to_sync.push(krate);
+        //        }
```
Comment (Jake-Shadle) [inline src/sync.rs]: Was this for testing?

File: src/sync.rs [hunk 2]
```diff
@@ -94,7 +94,7 @@ pub fn locked_crates(ctx: &crate::Ctx<'_>, root_dir: &Path) -> Result<(), Error>
 
     if to_sync.is_empty() {
         info!("all crates already available on local disk");
-        return Ok(());
+        //        return Ok(());
```
Comment (Jake-Shadle) [inline src/sync.rs]: DItto

**Comment by Jake-Shadle** [comment]
Also, it's ok to remove the openssl entries in the deny.toml since s3 is now behind a feature gate, so people can opt out of it if they don't want s3 support.

**Comment by Jake-Shadle** [comment]
I'm going to go ahead and merge this and so some fixup, and looking at rusoto, I think I can use a feature to use rustls instead of openssl which is good.

**Comment by Jake-Shadle** [comment]
Thanks very much for the PR!

**Comment by repi** [comment]
Should add the new S3 feature to the README also

**Comment by Jake-Shadle** [comment]
Yup, I'll do that after a bit of refactoring.

**Pull request merged**